### PR TITLE
Use e1000 ethernet driver in qemu_x86

### PR DESCRIPTION
**Pull request by jukkar**
Allow user transparently use e1000 ethernet driver in qemu_x86 board.
Usage:
```
cd samples/net/echo_server
mkdir build && cd build
cmake .. -DBOARD=qemu_x86 -DOVERLAY_CONFIG=overlay-e1000.conf
make run
```

You must also run `net-setup.sh` script from net-tools project that will setup host side corretly.

There is some configuration issue either with e1000 or in qemu as the connectivity does not really work as expected. Some initial IPv6 packets are seen in the interface when zephyr boots but after few seconds no packets are received nor sent by zephyr.

### CODE REVIEW

**Comment by jukkar** [comment]
> Should this be done before or after "make run"?

Before, so that host setup is ok and qemu can be started as a normal user.

> So, should this be DNM for now?

There is DNM label set already for this PR.

**Review by ozhuraki — APPROVED** [review]
LGTM

**Comment by pfalcon** [comment]
> There is some configuration issue either with e1000 or in qemu as the connectivity does not really work as expected. Some initial IPv6 packets are seen in the interface when zephyr boots but after few seconds no packets are received nor sent by zephyr.

Ok, so for me the situation remains the same as before (as was reported in https://github.com/zephyrproject-rtos/zephyr/pull/9112): Zephyr can send packets to host. Host receives them. Host sends replies back. But Zephyr never receives anything, and reason for that is quite obvious: e1000 RX interrupt never fires!

Here's for example how `net ping 192.0.2.2` from Zephyr shell looks in Wireshark:

~~~
10	50.580612214	RealtekU_12:34:56	Broadcast	ARP	42		Who has 192.0.2.2? Tell 0.0.0.0
11	50.580622235	IcannIan_00:53:ff	RealtekU_12:34:56	ARP	42		192.0.2.2 is at 00:00:5e:00:53:ff
~~~

I.e. Z sends ARP, the host replies, but Z never see that reply:

~~~
uart:~$ net stats

Interface 0x00426660 (Ethernet) [0]
===================================
IPv6 recv      0	sent	5	drop	0	forwarded	0
IPv6 ND recv   0	sent	5	drop	0
IPv6 MLD recv  0	sent	0	drop	0
IPv4 recv      0	sent	2	drop	0	forwarded	0
IP vhlerr      0	hblener	0	lblener	0
IP fragerr     0	chkerr	0	protoer	0
ICMP recv      0	sent	2	drop	0
ICMP typeer    0	chkerr	0
UDP recv       0	sent	0	drop	0
UDP chkerr     0
TCP bytes recv 0	sent	0
TCP seg recv   0	sent	0	drop	0
TCP seg resent 0	chkerr	0	ackerr	0
TCP seg rsterr 0	rst	0	re-xmit	0
TCP conn drop  0	connrst	0
Bytes received 0
Bytes sent     344
Processing err 0
~~~

Similarly for IPv6:

~~~
15	153.520130520	2001:db8::1	ff02::1:ff00:2	ICMPv6	86		Neighbor Solicitation for 2001:db8::2 from 52:54:00:12:34:56
16	153.520165735	2001:db8::2	2001:db8::1	ICMPv6	86		Neighbor Advertisement 2001:db8::2 (sol, ovr) is at 00:00:5e:00:53:ff
~~~


@jukkar, we have many cases when such investigations were deadlocked because another side was too ~~lazy~~ busy to do the obvious next steps in testing. Let's try to get on the same line now:

1. Please apply the following patch:
~~~
diff --git a/drivers/ethernet/eth_e1000.c b/drivers/ethernet/eth_e1000.c
index 996d4b12d8..d783ff72b7 100644
--- a/drivers/ethernet/eth_e1000.c
+++ b/drivers/ethernet/eth_e1000.c
@@ -130,6 +130,7 @@ out:
 
 static void e1000_isr(struct device *device)
 {
+printk("e1000_isr\n");
        struct e1000_dev *dev = device->driver_data;
        u32_t icr = ior32(dev, ICR); /* Cleared upon read */
~~~
2. Please show the exact dump from "make run" followed `net ping 192.0.2.2`, with either "e1000_isr" being printed, or with none of them.

For the output is:
~~~
SeaBIOS (version rel-1.11.2-0-gf9626ccb91-prebuilt.qemu-project.org)

iPXE v1.0.0-591-g7aee315
iPXE (http://ipxe.org) 00:02.0 C000 PCI2.10 PnP PMM+007983C0+007583C0 C000
                                                                               

Booting from ROM..

uart:~$ ***** Booting Zephyr OS zephyr-v1.13.0-1622-g8aa65494c8 *****

uart:~$ net ping 192.0.2.2
Sent a ping to 192.0.2.2
Ping timeout
uart:~$ QEMU: Terminated
~~~

**Comment by pfalcon** [comment]
Oh, and I'm on:
~~~
ZEPHYR_SDK_INSTALL_DIR=/home/pfalcon/opt/zephyr-sdk-0.9.5
~~~
of course

File: drivers/net/Kconfig [hunk 1]
```diff
@@ -35,6 +35,8 @@
 #
 menuconfig SLIP
 	bool "SLIP driver"
+	default y if NET_QEMU_SLIP
```
Comment (SebastianBoe) [inline drivers/net/Kconfig]: SLIP has a prompt.

This means it is user-settable and the value can be configured to be different from the default.

Is it valid to have NET_QEMU_SLIP, but not SLIP?

If not you need to either get rid of the prompt, or use select, or use depends.

**Comment by jukkar** [comment]
@pfalcon Please note that this PR is not really about e1000 and if it works or not. The motivation about this PR was to make integration to user workflow seamless, this PR could probably easily used with stellaris driver too.

As I said also in description, the connectivity does not really work for me either. I see some packets sent / received for few seconds but after that the interface becomes a black hole. Of course the e1000 needs to be get working before this PR can be merged, it would be very confusing for end user if this is merged with broken e1000 driver.

File: drivers/net/Kconfig (hunk 1 above)
Comment (jukkar) [reply drivers/net/Kconfig]: The NET_QEMU_SLIP requires SLIP. It is possible to use SLIP but not use the networking part like we do now.

**Comment by pfalcon** [comment]
@jukkar 

> @pfalcon Please note that this PR is not really about e1000 and if it works or not.

So, do you suggest that I should resend this request:

> Please show the exact dump from "make run" followed net ping 192.0.2.2, with either "e1000_isr" being printed, or with none of them.

to you via private email or some other means, or would you be able to reply here after all?

**Comment by pfalcon** [comment]
@jukkar: Oh, that's easy to rectify. Can you please do the following:

1. Please apply the following patch:
~~~
diff --git a/drivers/ethernet/eth_e1000.c b/drivers/ethernet/eth_e1000.c
index 996d4b12d8..d783ff72b7 100644
--- a/drivers/ethernet/eth_e1000.c
+++ b/drivers/ethernet/eth_e1000.c
@@ -130,6 +130,7 @@ out:
 
 static void e1000_isr(struct device *device)
 {
+printk("e1000_isr\n");
        struct e1000_dev *dev = device->driver_data;
        u32_t icr = ior32(dev, ICR); /* Cleared upon read */
~~~
2. Please show the exact dump from echo_server's "make run" followed `net ping 192.0.2.2`, with either "e1000_isr" being printed, or with none of them.

**Comment by jukkar** [comment]
>     1. Please show the exact dump from echo_server's "make run" followed `net ping 192.0.2.2`, with either "e1000_isr" being printed, or with none of them.

I do not see any prints either
```
SeaBIOS (version rel-1.11.2-0-gf9626ccb91-prebuilt.qemu-project.org)

iPXE v1.0.0-591-g7aee315
iPXE (http://ipxe.org) 00:02.0 C000 PCI2.10 PnP PMM+007983C0+007583C0 C000
                                                                               

Booting from ROM..

uart:~$ ***** Booting Zephyr OS zephyr-v1.13.0-1632-g2f3315601b *****
uart:~$ net ping 192.0.2.2
Sent a ping to 192.0.2.2
Ping timeout
uart:~$ 
```

**Comment by jukkar** [comment]
New version where I refactored the SLIP Kconfig file.

**Comment by pfalcon** [comment]
> I do not see any prints either

Thanks! So after all, we're on the same line here with the behavior we observe.

**Comment by SebastianBoe** [comment]
> New version where I refactored the SLIP Kconfig file.

Could you see if it is possible to reduce the code duplication?

Does

depends on (! QEMU_TARGET || NET_QEMU_SLIP)

work, for instance?

If not then perhaps @ulfalizer knows the cleanest way to solve this.

**Comment by jukkar** [comment]
> Does
> 
> depends on (! QEMU_TARGET || NET_QEMU_SLIP)
> 
> work, for instance?

Your suggestion works just fine, thanks for the hint.

**Comment by jukkar** [comment]
Update:
* SLIP Kconfig refactoring
* NIC model selection made generic so it can be used with any qemu

**Review by pfalcon — APPROVED** [review]
So, I think that this patch is good and lays good foundation for supporting qemu ethernet emulation in Zephyr. It's unfortunate that the only qemu ethernet driver in the mainline is broken, but having *these* changes in the mainline would help with developing other qemu ethernet drivers, and may help with establishing best practices on how to handle qemu startup, and would help avoiding duplicate effort in that regard.

So, I'm voting for removing DNM label and merging this soon.

File: cmake/app/boilerplate.cmake [hunk 2]
```diff
@@ -327,6 +327,30 @@ message(STATUS "Cache files will be written to: ${USER_CACHE_DIR}")
 
 include(${BOARD_DIR}/board.cmake OPTIONAL)
 
+# If we are using the suitable ethernet driver inside qemu, then
```
Comment (pfalcon) [inline cmake/app/boilerplate.cmake]: Just testing results of my continued attention to the English articles usage. So, both in "the suitable ethernet driver" and "the zephyr instance", the "a" article should be used instead. Feel free to ignore unless a native speaker confirms and insists on changes ;-).

**Comment by jukkar** [comment]
New version:
* fix bad English
* simplify qemu command line option, -nic option can be used directly in our qemu version so do that

**Review by SebastianBoe — APPROVED** [review]

**Review by tbursztyka — APPROVED** [review]

**Review by nashif — APPROVED** [review]

**Pull request merged**